source/common/config/delta_subscription_state.cc

```diff
@@ -214,26 +243,44 @@ DeltaSubscriptionState::getNextRequestWithAck(const UpdateAck& ack) {
 }
 
 void DeltaSubscriptionState::addResourceState(
-    const envoy::service::discovery::v3::Resource& resource) {
+    const envoy::service::discovery::v3::Resource& resource, ResourceType type) {
   if (resource.has_ttl()) {
     ttl_.add(std::chrono::milliseconds(DurationUtil::durationToMilliseconds(resource.ttl())),
              resource.name());
   } else {
     ttl_.clear(resource.name());
   }
 
-  resource_state_[resource.name()] = ResourceState(resource);
-  resource_names_.insert(resource.name());
+  if (auto it = resource_state_.find(resource.name()); it != resource_state_.end()) {
+    auto old_type = it->second.type();
+    it->second = ResourceState(resource, effectiveResourceType(old_type, type));
+  } else {
+    resource_state_.insert({resource.name(), ResourceState(resource, type)});
+  }
+}
+
+bool DeltaSubscriptionState::setResourceWaitingForServer(const std::string& resource_name) {
```

Comment:
The semantics of this call is confusing (I don't think we follow this convention anywhere else). Please either change the name to indicate a possibility that resource state won't be set, or consider breaking the method up.

Reply:
I changed it to `OptRef<ResourceState> getResourceState(resource_name)` and added a modifier to ResourceState to reset it to the "waiting for server mode".